GitUI/CommandsDialogs/SettingsDialog/Pages/ColorsSettingsPage.cs

```diff
@@ -171,96 +189,98 @@ private void ChkHighlightAuthored_CheckedChanged(object sender, EventArgs e)
+            {
+                return;
+            }
+
+            BeginUpdateThemeSettings();
+            UseSystemVisualStyle = SelectedThemeId == ThemeId.Default;
+            EndUpdateThemeSettings();
         }
 
-        private void BtnResetTheme_Click(object sender, EventArgs e)
+        private void ChkUseSystemVisualStyle_CheckedChanged(object sender, EventArgs e)
         {
-            ThemeModule.Controller.ResetTheme();
-            chkUseSystemVisualStyle.Checked = true;
+            BeginUpdateThemeSettings();
+            EndUpdateThemeSettings();
         }
 
-        private void ComboBoxTheme_SelectedIndexChanged(object sender, EventArgs e)
+        private void BeginUpdateThemeSettings()
         {
-            var menu = (ComboBox)sender;
-            if (menu.SelectedIndex < 0)
-            {
-                return;
-            }
+            _updateThemeSettingsCounter++;
+        }
 
-            if (_syncingTheme)
-            {
-                return;
-            }
+        private bool UpdatingThemeSettings() =>
+            _updateThemeSettingsCounter > 0;
 
-            _syncingTheme = true;
-            try
+        private void EndUpdateThemeSettings()
+        {
+            int counter = --_updateThemeSettingsCounter;
+            if (counter < 0)
             {
-                ThemeModule.Controller.SetTheme((string)menu.SelectedItem);
+                throw new InvalidOperationException($"{nameof(EndUpdateThemeSettings)} must be called after {nameof(BeginUpdateThemeSettings)}");
             }
-            finally
+
+            if (counter == 0)
             {
-                _syncingTheme = false;
+                bool settingsChanged =
+                    UseSystemVisualStyle != ThemeModule.Settings.UseSystemVisualStyle ||
+                    SelectedThemeId != ThemeModule.Settings.Theme.Id;
+
+                lblRestartNeeded.Visible = settingsChanged;
+                chkUseSystemVisualStyle.Enabled = SelectedThemeId != ThemeId.Default;
             }
         }
 
-        private void ChkUseSystemVisualStyle_CheckedChanged(object sender, EventArgs e)
+        private struct FormattedThemeId
         {
-            ThemeModule.Controller.UseSystemVisualStyle = chkUseSystemVisualStyle.Checked;
-            UpdateRestartWarningVisibility();
-        }
+            private static readonly TranslationString FormatBuiltinThemeName =
+                new TranslationString("{0}");
 
-        private void Theme_Changed(bool colorsChanged, string themeName)
-        {
-            if (colorsChanged)
-            {
-                UpdateAppColors();
-            }
+            private static readonly TranslationString FormatUserDefinedThemeName =
+                new TranslationString("{0}, user-defined");
 
-            if (!string.IsNullOrEmpty(themeName))
+            private static readonly TranslationString DefaultThemeName =
+                new TranslationString("default");
```

Comment:
I don't think this will work

Reply:
moved `TranslationString` definitions to parent class `ColorSettingsPage`